pkg/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,391 @@
+			} else {
+				klog.ErrorS(err, "Failed to release IP", "ip", allocation.ip, "pool", allocation.ipPool)
+				return err
+			}
+		} else {
+			klog.InfoS("Released external IP", "servuce", service, "ip", allocation.ip, "pool", allocation.ipPool)
+		}
+		delete(c.ipAllocationMap, service)
+	}
+	return nil
+}
+
+func (c *ServiceExternalIPController) getExternalIPAllocation(service apimachinerytypes.NamespacedName) (*ipAllocation, bool) {
+	c.ipAllocationMutex.RLock()
+	defer c.ipAllocationMutex.RUnlock()
+	allocation, exist := c.ipAllocationMap[service]
+	return allocation, exist
+}
+
+func getServiceExternalIP(service *corev1.Service) string {
+	if len(service.Status.LoadBalancer.Ingress) == 0 {
+		return ""
+	}
+	return service.Status.LoadBalancer.Ingress[0].IP
+}
+
+func (c *ServiceExternalIPController) syncService(key apimachinerytypes.NamespacedName) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Service for %s. (%v)", key, time.Since(startTime))
+	}()
+
+	service, err := c.serviceLister.Services(key.Namespace).Get(key.Name)
+	if err != nil {
+		// Service already deleted
+		if apimachineryerrors.IsNotFound(err) {
+			if err := c.releaseExternalIP(key); err != nil {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+
+	// Service does not need external IP or type has changed.
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		if err := c.releaseExternalIP(key); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	currentIPPool := service.ObjectMeta.Annotations[antreaagenttypes.ServiceExternalIPPoolAnnotationKey]
+	prevIPAllocation, allocationExists := c.getExternalIPAllocation(key)
+	currentExternalIP := getServiceExternalIP(service)
+
+	// If user specifies external IP in spec, we should check whether it matches the current external IP.
+	specIPMatched := service.Spec.LoadBalancerIP == "" || service.Spec.LoadBalancerIP == currentExternalIP
+
+	if allocationExists && specIPMatched &&
+		c.externalIPAllocator.IPPoolExists(currentIPPool) &&
+		c.externalIPAllocator.IPPoolHasIP(currentIPPool, prevIPAllocation.ip) &&
+		currentIPPool == prevIPAllocation.ipPool &&
+		currentExternalIP == prevIPAllocation.ip.String() {
+		return nil
+	}
+
+	// The ExternalIPPool does not exist or has been deleted. Reclaim the external IP.
+	if currentIPPool != "" && !c.externalIPAllocator.IPPoolExists(currentIPPool) {
+		if err := c.releaseExternalIP(key); err != nil {
+			return err
+		}
+		if currentExternalIP != "" {
+			toUpdate := service.DeepCopy()
+			toUpdate.Status.LoadBalancer.Ingress = nil
+			return c.updateServiceStatus(service, toUpdate)
+		}
+	}
+
+	// the external IP or ExternalIPPool changed somehow. Delete the previous allocation.
+	if err := c.releaseExternalIP(key); err != nil {
+		return err
+	}
+
+	if currentIPPool == "" {
+		klog.ErrorS(fmt.Errorf("annotation %q is required", antreaagenttypes.ServiceExternalIPPoolAnnotationKey),
```

Comment:
I feel this should be Info and just say the service is ignored because no required annotation is found, otherwise it would keep logging error when users don't want to use Antrea to manage external IP, unless we plan to introduce a toggle for this controller.

Reply:
ok. changed to info level logging.